### ISSUE TITLE
NoInlineFullyQualifiedClassName vs UnusedUseStatementSniff.php

### ISSUE THREAD

**Issue by gmponos** [php-fig-rectified/psr2r-sniffer#9]
Lets say that I have the following

```php
<?php
class ClassA {

    public function __construct(PimpleContainer $container){
         /** \MyNamespace\ClassB $classB */
         $classB = $container['class_b'];
    }
}
```

If I understood correctly the above is not allowed by the `NoInlineFullyQualifiedClassName`

But also the below is not allowed by the `UnusedUseStatement`

```php
<?php
use \MyNamespace\ClassB;

class ClassA {

    public function __construct(PimpleContainer $container){
         /** ClassB $classB */
         $classB = $container['class_b'];
    }
}
```

**Comment by dereuromark** [comment]
Hi!

    /** \MyNamespace\ClassB $classB */

should be fine

but right now it still does

    /* \MyNamespace\ClassB $classB */

I would like to fix this. If you can help here, that would also be great.
The correct sniff already exists here: https://github.com/spryker/code-sniffer

**Comment by dereuromark** [comment]
Are you interested in migrating the sniff so all inline doc blocks are properly respected as such?

**Comment by dereuromark** [comment]
So the /** part is fixed.
But you are right, the unused use does not detect the use instead of inline doc blocks.

We should add that.

**Comment by dereuromark** [comment]
Actually, `/** @var ClassB $classB */` is the correct one (you missed the @ var part).
Then it works as expected, and the sniffer makes this to:

<?php

class ClassA {

	public function __construct(PimpleContainer $container){
		/** @var \MyNamespace\ClassB $classB */
		$classB = $container['class_b'];
	}

}

**Comment by gmponos** [comment]
I am so sorry about this. I think I explained this wrong from the start.

My initial goal was to use the `UnusedUseStatementSniff`. In order to test it in my projects I copy-pasted the code and saw that it was throwing a warning for the following cases:

```php
// It finds this as unused
// Even if I use /** or /* same thing.
use \MyNamespace\ClassB; 

class ClassA {

    public function __construct(PimpleContainer $container){
         /** @var ClassB $classB */ 
         $classB = $container['class_b'];
    }
}
```

Then I wonder around the folder and found out about the `NoInlineFullyQualifiedClassName`. I didn't test it but I thought that there might be a logical error here.

If no InlineFQCN is allowed and the use sniff does not find the uses then something will break. Why because if you use fixer then this will happen.


```php
class ClassA {

    public function __construct(PimpleContainer $container){
         /** @var ClassB $classB */ 
         $classB = $container['class_b'];
    }
}
```
The use was removed and I can't use inlineFQCN.

> Are you interested in migrating the sniff so all inline doc blocks are properly respected as such?

Anyway I would be happy to help but I am not sure how devoted can I be because of (job etc). Can you give me an example what you need me to fix?

> Actually, /** @var ClassB $classB */ is the correct one (you missed the @ var part).
Then it works as expected, and the sniffer makes this to:

Most probably this was missed at my examples.. my bad.. but it was there in the code that I tested your sniffs.

**Comment by gmponos** [comment]
After I have tested also the InlineFQCN sniff I saw that it does not check for the inlinde DocBlocks about FQCN statements.

Please have the scenario above in your mind if you ever add the inline DocBlocks in that sniff.

**Comment by gmponos** [comment]
> https://github.com/php-fig-rectified/psr2r-sniffer/issues/9#issuecomment-277529463

I guess after [this](https://github.com/php-fig-rectified/psr2r-sniffer/pull/7/commits/60f889fcb42fce0e06d999e39fc620429faa3623) it will be fine

**Comment by dereuromark** [comment]
You are right, it should skip the removal until the other sniff actually made it "really" not needed.
Thanks for clarifying it.
Are you able to PR a fix here?

**Comment by gmponos** [comment]
I am not sure I can make this work. Furthermore have you seen a related PR in codesniffer with title: _Add UnusedUseStatementSniff copied from Drupal Coder_ 

Check the last comment there about annotated `uses`.

Your psr2r-sniffer has a larger target group than the one on spryker where you use the same sniff about unused `uses`. It might need a lot of effort and cover more and more complex scenarios even with other sniffs that you have in order everyone to be happy.

I have never written a test for a Sniff so far. I am migrating my sniff to v3 of codesniffer hoping testing there will be easier. So I would prefer something smaller to help.

My opinion is that this sniff should be the last one added. What do you believe?

**Comment by ewbarnard** [comment]
@dereuromark I've converted NoInlineFullyQualifiedClassNameSniff to work with 3.x sniffer. However, it fails with this test case:
```
class NoInlineFullyQualifiedClassNameUnitTestInc extends \PSR2R\Base\AbstractBase {
	protected function getErrorList(\PHP_CodeSniffer\Util\Tokens $tokens = null) {
		return [
		];
	}

	protected function getWarningList(\PHP_CodeSniffer\Files\File $file = null) {
		return [
		];
	}
}
```

I believe the what's happening is this: When adding the second and later use statements, they get added (appended) to the same token. But, they get added in different fixer->beginChangeset() sequences. Fixer detects the same token being modified by multiple changesets, and assumes that sniffs are conflicting with each other, and the result is a mess.

Am I on the right track? Are there other sniffs which do multiple inserts in multiple changesets? I don't see any way to register a cleanup method in the sniff. If it existed, I could make a list of use statements to insert, and insert them all at once.

I could perhaps register for processing closing brace, and on the first processing pass, make note of the last closing brace position, and spit out the use statements upon seeing that token. That seems ugly and inefficient, though.

**Comment by ewbarnard** [comment]
I posted my question to Stack Overflow: https://stackoverflow.com/questions/44641322/how-to-create-multiple-use-statements-in-sniff

**Comment by ewbarnard** [comment]
Come to think of it, I could note the position of the last token from my list of registered tokens, and when I process that token, run through the list (if any) of use statements to insert. I'll try that and see if it answers.

**Comment by dereuromark** [comment]
Is this still open?

**Comment by dereuromark** [comment]
I can reproduce it for sure
It looks like the UnusedUse clears it out because at the time it thinks those are unneeded...

@ewbarnard What is the sentinel thing you added? Is there any test case I can run for it?
Thx

**Comment by dereuromark** [comment]
Can we make a PR here? What is the suggested fix?

**Comment by ewbarnard** [comment]
Hi Mark, looks like I missed your prior comment two years ago. Let me dig, see if I have a fix, and if not we should probably close it. Back to you later today.

**Comment by ewbarnard** [comment]
@dereuromark Everything I have was merged 2017-06-26 as pull request #10 - we should be able to close this issue, I think.
